In [ ]:
# !lsb_release -d

In [ ]:
# !free -m

In [ ]:
# !nvidia-smi

In [ ]:
!pip install sentence-transformers
!pip install sinling

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 86.0/86.0 kB 2.9 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 40.0 MB/s eta 0:00:00
  Created wheel for sentence-transformers: filename=sentence_transformers-2.2.2-py3-none-any.whl size=125923 sha256=fe2611c53eeebc937f7acd2b67fc69c56c158a60e444d48d30008428f8706651
  Stored in directory: /root/.cache/pip/wheels/62/f2/10/1e606fd5f02395388f74e7462910fe851042f97238cbbd902f
Successfully built sentence-transformers
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 20.0/20.0 MB 17.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 397.5/397.5 kB 42.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 993.5/993.5 kB 53.6 MB/s eta 0:00:00


In [ ]:
import io
import os
import time
import torch
import pickle
import collections
import numpy as np
from sinling import word_splitter
from sentence_transformers.util import cos_sim
from sentence_transformers import SentenceTransformer

In [ ]:
from google.colab import drive

drive.mount('/content/gdrive/', force_remount=True)

Mounted at /content/gdrive/


In [ ]:
!ls /content/gdrive/MyDrive/Multilingual_Embedding_Alignment/data

cc.en.300.vec
cc.en.300.vec_words.txt
cc_en-si_testset_1500_5000_200k.txt
cc_en-si_trainset_1500_5000_200k.txt
cc.si.300.vec
cc.si.300.vec_words.txt
cc_si-en_testset_1500_5000_200k.txt
cc_si-en_trainset_1500_5000_200k.txt
data
de-en.0-5000.txt
de-en.5000-6500.txt
de-en.txt
distiluse-base-multilingual-cased_wiki_embed_en_200000.pt
distiluse-base-multilingual-cased_wiki_embed_es_200000.pt
distiluse-base-multilingual-cased_wiki_embed_zh_200000.pt
en-de.0-5000.txt
en-de.5000-6500.txt
en-de.txt
en-es.0-5000.txt
en-es.5000-6500.txt
en-es.txt
en-fr.0-5000.txt
en-fr.5000-6500.txt
en-fr.txt
en-it.0-5000.txt
en-it.5000-6500.txt
en-it.txt
en-ja.0-5000.txt
en-ja.5000-6500.txt
en-ja.txt
en-ru.0-5000.txt
en-ru.5000-6500.txt
en-ru.txt
en-ta.0-5000.txt
en-ta.5000-6500.txt
en-ta.txt
en-tr.0-5000.txt
en-tr.5000-6500.txt
en-tr.txt
en-zh.0-5000.txt
en-zh.5000-6500.txt
en-zh.txt
es-en.0-5000.txt
es-en.5000-6500.txt
es-en.txt
evaluate_labse_on_fasttext_vocabulary.ipynb
extract_fasttext_words.py
fr-en.0-5000

In [ ]:
!ln -s /content/gdrive/MyDrive/Multilingual_Embedding_Alignment/data data
!ls data

cc.en.300.vec
cc.en.300.vec_words.txt
cc_en-si_testset_1500_5000_200k.txt
cc_en-si_trainset_1500_5000_200k.txt
cc.si.300.vec
cc.si.300.vec_words.txt
cc_si-en_testset_1500_5000_200k.txt
cc_si-en_trainset_1500_5000_200k.txt
data
de-en.0-5000.txt
de-en.5000-6500.txt
de-en.txt
distiluse-base-multilingual-cased_wiki_embed_en_200000.pt
distiluse-base-multilingual-cased_wiki_embed_es_200000.pt
distiluse-base-multilingual-cased_wiki_embed_zh_200000.pt
en-de.0-5000.txt
en-de.5000-6500.txt
en-de.txt
en-es.0-5000.txt
en-es.5000-6500.txt
en-es.txt
en-fr.0-5000.txt
en-fr.5000-6500.txt
en-fr.txt
en-it.0-5000.txt
en-it.5000-6500.txt
en-it.txt
en-ja.0-5000.txt
en-ja.5000-6500.txt
en-ja.txt
en-ru.0-5000.txt
en-ru.5000-6500.txt
en-ru.txt
en-ta.0-5000.txt
en-ta.5000-6500.txt
en-ta.txt
en-tr.0-5000.txt
en-tr.5000-6500.txt
en-tr.txt
en-zh.0-5000.txt
en-zh.5000-6500.txt
en-zh.txt
es-en.0-5000.txt
es-en.5000-6500.txt
es-en.txt
evaluate_labse_on_fasttext_vocabulary.ipynb
extract_fasttext_words.py
fr-en.0-5000

In [ ]:
# model_card = 'sentence-transformers/LaBSE'
# model_card = 'sentence-transformers/distiluse-base-multilingual-cased-v1' #'distiluse-base-multilingual-cased'
model_card = 'distiluse-base-multilingual-cased'
# model_card = 'microsoft/mpnet-base'
# model_card = 'sentence-transformers/stsb-xlm-r-multilingual'
model_name = model_card.rsplit('/')[-1]
model_name

'distiluse-base-multilingual-cased'

In [ ]:
print("[INFO] Loading the encoder..")
model = SentenceTransformer(model_card)
print("[INFO] Encoder loaded..")

[INFO] Loading the encoder..


.gitattributes:   0%|          | 0.00/690 [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

2_Dense/config.json:   0%|          | 0.00/114 [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/1.58M [00:00<?, ?B/s]

rust_model.ot:   0%|          | 0.00/1.58M [00:00<?, ?B/s]

README.md:   0%|          | 0.00/2.37k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/607 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/122 [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/539M [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.96M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/528 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/996k [00:00<?, ?B/s]

modules.json:   0%|          | 0.00/341 [00:00<?, ?B/s]

[INFO] Encoder loaded..


In [ ]:
def load_vocab(file_path, maxload=None):
    with open(file_path) as f:
        vocab = [x.strip() for x in f.readlines()]

    n_vocab = len(vocab)

    if maxload is None:
        maxload = n_vocab

    print(f"[INFO] Found {len(vocab)} entries and, returning {min(maxload, n_vocab)} entries...")

    return vocab[:maxload]

In [ ]:
def compute_embeddings(sentences):
    t0 = time.time()
    embeddings = model.encode(sentences, convert_to_tensor=True)
    print(f"[INFO] Encoding time {(time.time() - t0) * 1000} ms")

    print(embeddings.shape, embeddings[0].shape)

    return embeddings

In [ ]:
def wi(words):
    w2i = {}
    i2w = {}
    for i, w in enumerate(words):
        if w not in w2i:
            w2i[w] = i
            i2w[i] = w

    return w2i, i2w

In [ ]:
def idx(words):
    w2i = {}
    for i, w in enumerate(words):
        if w not in w2i:
            w2i[w] = i

    return w2i


def load_lexicon(filename, words_src, words_tgt, verbose=True):
    f = io.open(filename, 'r', encoding='utf-8')
    lexicon = collections.defaultdict(set)
    idx_src , idx_tgt = idx(words_src), idx(words_tgt)
    vocab = set()

    for line in f:
        # print(line)
        l_split = line.split()
        word_src, word_tgt = l_split[0], " ".join(l_split[1:])
        if word_src in idx_src and word_tgt in idx_tgt:
            lexicon[idx_src[word_src]].add(idx_tgt[word_tgt])
            # print(f"[DEBUG] added {word_src}, {word_tgt}")
        vocab.add(word_src)

    if verbose:
        coverage = len(lexicon) / float(len(vocab))
        print("[INFO] Coverage of source vocab: %.4f" % (coverage))

    return lexicon, float(len(vocab))

In [ ]:
# si_splits_file_path = f'data/si_splits_{}.pkl'

# if not os.path.exists(si_splits_file_path):
#     with open(si_splits_file_path, 'wb') as f:
#         pickle.dump({}, f)

#     si_splits = {}

# else:
#     with open(si_splits_file_path, 'rb') as f:
#         si_splits = pickle.load(f)

In [ ]:
# !mv data/si_splits.pkl data/si_splits_corruped.pkl

In [ ]:
# si_splits

In [ ]:
def lematized_match(ref, pred):
    # ref = lexicon[idx_src[j]]
    # pred = pred[j - i]

    if not isinstance(ref, list):
        ref = list(ref)

    if not isinstance(pred, list):
        pred = list(pred)

    # any(one_top_k in list(lexicon[idx_src[j]]) for one_top_k in pred[j - i])

    split_ref = []

    for x in ref:

      if isinstance(x, torch.Tensor):
        x = x.item()

      if x in si_splits:
        split_ref.append(si_splits[x])
      else:
        try:
          split = word_splitter.split(words_tgt[x])['base']
        except:
          continue

        try:
          split = idx_tgt[split]
        except:
          ...

        split_ref.append(split)
        si_splits[x] = split

        with open(si_splits_file_path, 'wb') as f:
            pickle.dump(si_splits, f)

    split_pred = []

    for y in pred:

      x = y.item()

      if x in si_splits:
        split_pred.append(si_splits[x])
      else:
        try:
          split = word_splitter.split(words_tgt[x])['base']
        except:
          continue

        try:
          split = idx_tgt[split]
        except:
          ...

        split_pred.append(split)
        si_splits[x] = split

        with open(si_splits_file_path, 'wb') as f:
            pickle.dump(si_splits, f)

    split_ref = ref + split_ref
    split_pred = pred + split_pred

    return any(one_top_k in split_ref for one_top_k in split_pred)


In [ ]:
def compute_nn_accuracy(x_src, x_tgt, lexicon, bsz=100, lexicon_size=-1, top_k=1, norm=False):  # use norm=False if vectors are already normalized
    if lexicon_size < 0:
        lexicon_size = len(lexicon)

    idx_src = list(lexicon.keys())
    acc = 0.0

    if norm:
        # x_src /= np.linalg.norm(x_src, axis=1)[:, np.newaxis] + 1e-8
        for i in range(len(x_src)):
            x_src[i] /= np.linalg.norm(x_src[i]) + 1e-8

    if norm:
        # x_tgt /= np.linalg.norm(x_tgt, axis=1)[:, np.newaxis] + 1e-8
        for i in range(len(x_tgt)):
            x_tgt[i] /= np.linalg.norm(x_tgt[i]) + 1e-8

    for i in range(0, len(idx_src), bsz):
        e = min(i + bsz, len(idx_src))
        scores = cos_sim(x_tgt, x_src[idx_src[i:e]])
        # print(scores.shape)

        # pred = scores.argsort(dim=0, descending=False)[-top_k:].T
        pred = scores.argsort(dim=0, descending=True)[:top_k].T
        # print(pred.shape)
        for k, j in enumerate(range(i, e)):
            # print(f"Pair: {words_src[idx_src[j]]} === {[words_tgt[tg] for tg in lexicon[idx_src[j]]]} ({lexicon[idx_src[j]]}) =================")

            if any(one_top_k in list(lexicon[idx_src[j]]) for one_top_k in pred[j - i]):
                acc += 1.0

            elif tgt == "si" and do_lematized_match and lematized_match(ref=list(lexicon[idx_src[j]]), pred=pred[j - i]):
                acc += 1.0

            # for one_top_k in pred[j - i]:
            #     print(f"{words_src[idx_src[j]]} -- {tgt_vocab[one_top_k]} ({one_top_k}) --> {scores.T[k][one_top_k]}")
            #     if one_top_k in list(lexicon[idx_src[j]]):
            #         acc += 1.0
            #         print(f"{one_top_k}*******************************************")
            #         break
        print(f"{i + bsz} done -- correct: {acc}")
        # if i > 0:
        #     break
    print(f"correct: {acc}")
    return acc / lexicon_size

In [ ]:
def compute_csls_accuracy(x_src, x_tgt, lexicon, lexicon_size=-1, k=10, bsz=1024, top_k=1, norm=False): #, norm=False, norm_src_only=False):
    if lexicon_size < 0:
        lexicon_size = len(lexicon)

    idx_src = list(lexicon.keys())

    if norm:
      # x_src /= np.linalg.norm(x_src, axis=1)[:, np.newaxis] + 1e-8
      for i in range(len(x_src)):
          x_src[i] /= np.linalg.norm(x_src[i]) + 1e-8

    if norm:
      # x_tgt /= np.linalg.norm(x_tgt, axis=1)[:, np.newaxis] + 1e-8
      for i in range(len(x_tgt)):
          x_tgt[i] /= np.linalg.norm(x_tgt[i]) + 1e-8

    sr = x_src[list(idx_src)]
    # sc = np.dot(sr, x_tgt.T)
    sc = torch.matmul(sr, x_tgt.T)
    similarities = 2 * sc
    # sc2 = np.zeros(x_tgt.shape[0])
    sc2 = torch.zeros(x_tgt.shape[0]).to(similarities.device)

    for i in range(0, x_tgt.shape[0], bsz):
        j = min(i + bsz, x_tgt.shape[0])
        # sc_batch = np.dot(x_tgt[i:j, :], x_src.T)
        sc_batch = torch.matmul(x_tgt[i:j, :], x_src.T)
        # dotprod = np.partition(sc_batch, -k, axis=1)[:, -k:]
        dotprod = torch.topk(sc_batch, k=k, dim=1).values
        # sc2[i:j] = np.mean(dotprod, axis=1)
        sc2[i:j] = torch.mean(dotprod, dim=1)
        torch.cuda.empty_cache()

        # print(f"{i + bsz} done")
    torch.cuda.empty_cache()
    # similarities -= sc2[np.newaxis, :]
    similarities -= torch.unsqueeze(sc2, dim=0)

    del sc_batch, dotprod, sc2
    torch.cuda.empty_cache()

    nn = []
    # # nn = np.argmax(similarities, axis=1).tolist()
    # # nn = np.argsort(similarities, axis=1).T[-top_k:][::-1].T
    # nn = torch.argsort(similarities, axis=1).T[-top_k:].flipud().T
    for sim_vec in similarities:
      nn.append(torch.argsort(sim_vec, dim=-1)[-top_k:].flipud())
    # print(nn)
    correct = 0.0

    for k in range(0, len(lexicon)):

        # if nn[k] in lexicon[idx_src[k]]:
        if any(one_top_k.item() in lexicon[idx_src[k]] for one_top_k in nn[k]):
            correct += 1.0

        elif tgt == "si" and do_lematized_match and lematized_match(ref=lexicon[idx_src[k]], pred=nn[k]):
            correct += 1.0

    return correct / lexicon_size

In [ ]:
def find_top_k_for_src_word(test_word, target_word, x_src, x_tgt, idx_src, words_src, idx_tgt, words_tgt, top_k=5):
    if isinstance(test_word, str):
        src_embd = x_src[idx_src[test_word]]
        word_str = test_word
    else:
        src_embd = x_src[test_word]
        word_str = words_src[test_word]

    if isinstance(target_word, str):
        tgt_word_idx = idx_tgt[target_word]
        tgt_word_str = target_word
    else:
        tgt_word_str = words_tgt[target_word]
        tgt_word_idx = target_word

    scores = cos_sim(x_tgt, src_embd)

    print(scores.shape)

    # pred = scores.argsort(dim=0, descending=False)[-top_k:].T
    pred = scores.argsort(dim=0, descending=True)[:top_k].T
    print(pred, pred.shape)

    for one_top_k in pred[0]:
        one_top_k = one_top_k.item()
        print(f"{word_str} -- {words_tgt[one_top_k]} --> {scores.T[0][one_top_k]}")

    print(f"\n{word_str} -- {tgt_word_str} --> {scores.T[0][tgt_word_idx]}================")




In [ ]:
def evaluate_dataset_pairs(lexicon, idx_src, words_src, idx_tgt, words_tgt):
    t0 = time.time()

    for i, (src_id, tgt_ids) in enumerate(lexicon.items()):
        src_word = words_src[src_id]
        tgt_words = [words_tgt[tgt_id] for tgt_id in tgt_ids]

        print("---" * 20)
        print(f"{src_word} == {tgt_words}")

        src_embd = model.encode(src_word, convert_to_tensor=True)
        # src_embd2 = x_src[src_id]
        tgt_embds = model.encode(tgt_words, convert_to_tensor=True)
        # tgt_embds2 = x_tgt[list(tgt_ids)]

        scores = cos_sim(src_embd, tgt_embds)
        # scores2 = cos_sim(src_embd2, tgt_embds2)
        # print(scores, scores2)
        for j in range(len(tgt_words)):
            print(f"{src_word} -- {tgt_words[j]} --> {scores[0][j]}")

        if i > 5:
            break


In [ ]:
dataset = "wiki"
# dataset = "cc"

# src = "en"
# src = "ta"
# src = "ja"
src = "tr"
# src = "it"
# src = "de"
# src = "fr"
# src = "es"
# src = "zh"
# src = "ru"
# src = "si"

tgt = "en"
# tgt = "tr"
# tgt = "ja"
# tgt = "it"
# tgt = "de"
# tgt = "ru"
# tgt = "si"
# tgt = "ta"
# tgt = "es"
# tgt = "fr"
# tgt = "zh"

# src = "si"
# tgt = "en"

maxload = 200_000
# maxload = 15000

src_prunable = (src=="si" or src=="zh" or src=="ru" or src=="ta" or src=="ja")
tgt_prunable = (tgt=="si" or tgt=="zh" or tgt=="ru" or tgt=="ta" or tgt=="ja")
# tgt_prunable = (tgt=="si" or tgt=="zh" or tgt=="ru" or tgt=="es")
do_lematized_match = True
# do_lematized_match = False
prune = True
# prune = False

In [ ]:
if tgt == "si" and do_lematized_match:
  si_splits_file_path = f'data/si_splits_{dataset}.pkl'

  if not os.path.exists(si_splits_file_path):
      with open(si_splits_file_path, 'wb') as f:
          pickle.dump({}, f)

      si_splits = {}

  else:
      with open(si_splits_file_path, 'rb') as f:
          si_splits = pickle.load(f)

  si_splits

In [ ]:
# len(si_splits)

In [ ]:
for lang in [src, tgt]:

  print(f"Language: {lang}")
  # os.environ["lang"] = "en"

  if dataset == "wiki":
    embed_model = f"wiki.{lang}.vec"

    if not os.path.isfile(f"data/{embed_model}"):
      print(f"[INFO] Downloading data/{embed_model}...")
      os.system(f'wget -c "https://dl.fbaipublicfiles.com/fasttext/vectors-wiki/{embed_model}" -P data/')

      print("[INFO] Dowinloading the MUSE alignment datasets...")
      os.system(f' wget -c "https://dl.fbaipublicfiles.com/arrival/dictionaries/{src}-{tgt}.5000-6500.txt" -P data')
      os.system(f' wget -c "https://dl.fbaipublicfiles.com/arrival/dictionaries/{src}-{tgt}.0-5000.txt" -P data')
      os.system(f' wget -c "https://dl.fbaipublicfiles.com/arrival/dictionaries/{src}-{tgt}.txt" -P data')

      os.system(f' wget -c "https://dl.fbaipublicfiles.com/arrival/dictionaries/{tgt}-{src}.5000-6500.txt" -P data')
      os.system(f' wget -c "https://dl.fbaipublicfiles.com/arrival/dictionaries/{tgt}-{src}.0-5000.txt" -P data')
      os.system(f' wget -c "https://dl.fbaipublicfiles.com/arrival/dictionaries/{tgt}-{src}.txt" -P data')

      print("[INFO] Extracting the vocabulary...")
      os.system(f'python data/extract_fasttext_words.py data/{embed_model}')
    else:
      print(f"[INFO] data/{embed_model} exists...")

  else:
    embed_model = f"cc.{lang}.300.vec"

    if not os.path.isfile(f"data/{embed_model}"):
      print(f"[INFO] Downloading data/{embed_model}...")
      os.system(f'wget -c "https://dl.fbaipublicfiles.com/fasttext/vectors-crawl/{embed_model}.gz" -P data/')
      os.system(f'gzip -dkv "data/{embed_model}.gz"')
      os.system(f'rm "data/{embed_model}.gz"')

      print("[INFO] Dowinloading the MUSE alignment datasets...")
      os.system(f' wget -c "https://dl.fbaipublicfiles.com/arrival/dictionaries/{src}-{tgt}.5000-6500.txt" -P data')
      os.system(f' wget -c "https://dl.fbaipublicfiles.com/arrival/dictionaries/{src}-{tgt}.0-5000.txt" -P data')
      os.system(f' wget -c "https://dl.fbaipublicfiles.com/arrival/dictionaries/{src}-{tgt}.txt" -P data')

      os.system(f' wget -c "https://dl.fbaipublicfiles.com/arrival/dictionaries/{tgt}-{src}.5000-6500.txt" -P data')
      os.system(f' wget -c "https://dl.fbaipublicfiles.com/arrival/dictionaries/{tgt}-{src}.0-5000.txt" -P data')
      os.system(f' wget -c "https://dl.fbaipublicfiles.com/arrival/dictionaries/{tgt}-{src}.txt" -P data')

      print("[INFO] Extracting the vocabulary...")
      os.system(f'python data/extract_fasttext_words.py data/{embed_model}')
    else:
      print(f"[INFO] data/{embed_model} exists...")

# !python extract_fasttext_words.py

Language: tr
[INFO] data/wiki.tr.vec exists...
Language: en
[INFO] data/wiki.en.vec exists...


In [ ]:
if dataset == "wiki":
    src_vocab_path = f"data/{dataset}.{src}.vec_words.txt"
    tgt_vocab_path = f"data/{dataset}.{tgt}.vec_words.txt"
else:
    src_vocab_path = f"data/{dataset}.{src}.300.vec_words.txt"
    tgt_vocab_path = f"data/{dataset}.{tgt}.300.vec_words.txt"

if src == "si" or tgt == "si":
  train_dataset_path = f"data/{dataset}_{src}-{tgt}_trainset_5000_5000_200k.txt"
  test_dataset_path = f"data/{dataset}_{src}-{tgt}_testset_1500_5000_200k.txt"
else:
  train_dataset_path = f"data/{src}-{tgt}.0-5000.txt"
  test_dataset_path = f"data/{src}-{tgt}.5000-6500.txt"

In [ ]:
def not_ASCII(word):
    is_ascii = [ord(x)<256 for x in word]

    if all(is_ascii):
        return False

    return True

In [ ]:
# x_src = compute_embeddings(src_vocab)
# torch.save(x_src, f"{model_name}_{dataset}_embed_{src}_{maxload}.pt")

In [ ]:
# x_tgt = compute_embeddings(tgt_vocab)
# torch.save(x_tgt, f"{model_name}_{dataset}_embed_{tgt}_{maxload}.pt")

In [ ]:
def load_saved_embeddings(embed_path, prune_ascii=False, vocab=None):

    if os.path.isfile(embed_path):
        print(f"[INFO] Loading the embeddings from {embed_path}...")
        embeds = torch.load(embed_path)
        print("[INFO] Loading done...")
    else:
        print(f"[INFO] {embed_path} does not exist. Calculating the embeddings..")
        embeds = compute_embeddings(vocab)
        torch.save(embeds, embed_path)
        print(f"[INFO] Saved the embeddings at {embed_path}...")

    print(f"[INFO] Original embeddings size: {embeds.shape}")

    if prune_ascii and vocab is not None:
        print("[INFO] Pruning the vocabulary by removing ASCII entries...")
        prune_idx = list(map(not_ASCII, vocab))
        embeds = embeds[prune_idx]
        vocab = list(np.array(vocab)[prune_idx])

        print(f"[INFO] After pruning embeddings size: {embeds.shape}")

    return embeds, vocab

In [ ]:
src_vocab = load_vocab(src_vocab_path, maxload=maxload)
tgt_vocab = load_vocab(tgt_vocab_path, maxload=maxload)

[INFO] Found 416051 entries and, returning 200000 entries...
[INFO] Found 2519370 entries and, returning 200000 entries...


In [ ]:
x_src, src_vocab = load_saved_embeddings(f"data/{model_name}_{dataset}_embed_{src}_{maxload}.pt", prune_ascii=prune and src_prunable, vocab=src_vocab)
x_tgt, tgt_vocab = load_saved_embeddings(f"data/{model_name}_{dataset}_embed_{tgt}_{maxload}.pt", prune_ascii=prune and tgt_prunable, vocab=tgt_vocab)

[INFO] Loading the embeddings from data/distiluse-base-multilingual-cased_wiki_embed_tr_200000.pt...
[INFO] Loading done...
[INFO] Original embeddings size: torch.Size([200000, 512])
[INFO] Loading the embeddings from data/distiluse-base-multilingual-cased_wiki_embed_en_200000.pt...
[INFO] Loading done...
[INFO] Original embeddings size: torch.Size([200000, 512])


In [ ]:
torch.cuda.empty_cache()

In [ ]:
# src_vocab
# tgt_vocab

In [ ]:
(idx_src, words_src) , (idx_tgt, words_tgt) = wi(src_vocab), wi(tgt_vocab)

In [ ]:
# words_tgt


In [ ]:
src2tgt, lexicon_size = load_lexicon(test_dataset_path, src_vocab, tgt_vocab)

[INFO] Coverage of source vocab: 1.0000


In [ ]:
evaluate_dataset_pairs(src2tgt, idx_src, words_src, idx_tgt, words_tgt)

------------------------------------------------------------
mitchell == ['mitchell']
mitchell -- mitchell --> 1.0
------------------------------------------------------------
başlıklar == ['banners', 'topics', 'titles', 'headlines', 'headers']
başlıklar -- banners --> 0.669366180896759
başlıklar -- topics --> 0.5946398377418518
başlıklar -- titles --> 0.7219560742378235
başlıklar -- headlines --> 0.720492959022522
başlıklar -- headers --> 0.6806329488754272
------------------------------------------------------------
büyü == ['grow', 'spell', 'magic']
büyü -- grow --> 0.3857605755329132
büyü -- spell --> 0.7243879437446594
büyü -- magic --> 0.9324667453765869
------------------------------------------------------------
palace == ['palace']
palace -- palace --> 1.0
------------------------------------------------------------
kırdı == ['broke']
kırdı -- broke --> 0.6269254684448242
------------------------------------------------------------
wolfgang == ['wolfgang']
wolfgang -- wolfgang

In [ ]:
# x_src = torch.load(f"{model_name}_{dataset}_embed_{src}_{maxload}.pt")
# x_tgt = torch.load(f"{model_name}_{dataset}_embed_{tgt}_{maxload}.pt")

In [ ]:
x_src.shape, x_tgt.shape

(torch.Size([200000, 512]), torch.Size([200000, 512]))

In [ ]:
lexicaon_keys = list(src2tgt.keys())
# lexicaon_keys

In [ ]:
test_word = lexicaon_keys[4]
target_word = list(src2tgt[test_word])[0]
find_top_k_for_src_word(test_word, target_word, x_src, x_tgt, idx_src, words_src, idx_tgt, words_tgt, top_k=10)

torch.Size([200000, 1])
tensor([[158051, 182739, 139704, 109672, 143065, 125432,  62862,  71004, 115570,
         156433]], device='cuda:0') torch.Size([1, 10])
kırdı -- rüştü --> 0.8553563952445984
kırdı -- riccò --> 0.8472211956977844
kırdı -- vrml --> 0.8394039869308472
kırdı -- diapause --> 0.832544207572937
kırdı -- issoufou --> 0.8202496767044067
kırdı -- murtala --> 0.8200274705886841
kırdı -- bezalel --> 0.8083859086036682
kırdı -- weixiu --> 0.803575873374939
kırdı -- nuchal --> 0.8016057014465332
kırdı -- kelmendi --> 0.7970110177993774

kırdı -- broke --> 0.626925528049469================


In [ ]:
# len(si_splits)

### Test Codes

In [ ]:
import numpy as np
import torch

In [ ]:
# in_arr = np.array([ [2, 0,  1, 5, 4, 9], [2, 0,  1, 5, 4, 9]])
in_arr = np.array([ [2, 0, 1, 5, 4, 9, 4, 3], [5, 0, 1, 2, 4, 9, 2, 3]])
print ("Input array : ", in_arr)

out_arr = np.partition(in_arr, -2, axis=1)
print ("Output partitioned array : ", out_arr)

# out_arr = np.partition(in_arr, -2, axis=1)
# print ("Output partitioned array : ", out_arr)

Input array :  [[2 0 1 5 4 9 4 3]
 [5 0 1 2 4 9 2 3]]
Output partitioned array :  [[0 2 1 3 4 4 5 9]
 [2 2 1 0 3 4 5 9]]


In [ ]:
in_tensor = torch.tensor(in_arr)
in_tensor

tensor([[2, 0, 1, 5, 4, 9, 4, 3],
        [5, 0, 1, 2, 4, 9, 2, 3]])

In [ ]:
out_tensor = torch.topk(in_tensor, k=2, dim=1)
print ("Output partitioned tensor : ", out_tensor.values)
torch.mean(out_tensor.values.float(), dim=1)

Output partitioned tensor :  tensor([[9, 5],
        [9, 5]])


tensor([7., 7.])

In [ ]:
a1 = np.argsort(in_arr, axis=1)
print(f"{a1=}")
a2 = a1.T[-3:]
print(f"{a2=}")
a3 = a2[::-1]
print(f"{a3=}")
a4 = a3.T
print(f"{a4=}")

a1=array([[1, 2, 0, 7, 4, 6, 3, 5],
       [1, 2, 3, 6, 7, 4, 0, 5]])
a2=array([[6, 4],
       [3, 0],
       [5, 5]])
a3=array([[5, 5],
       [3, 0],
       [6, 4]])
a4=array([[5, 3, 6],
       [5, 0, 4]])


In [ ]:
# nn_t = torch.argsort(in_tensor, dim=1).T[-3:][::-1].T
# nn_t = torch.argsort(in_tensor, dim=1).T[-3:].fliplr().T
t1 = torch.argsort(in_tensor, dim=1)
print(f"{t1=}")
t2 = t1.T[-3:]
print(f"{t2=}")
t3 = t2.flipud()
print(f"{t3=}")
t4 = t3.T
print(f"{t4=}")

t1=tensor([[1, 2, 0, 7, 4, 6, 3, 5],
        [1, 2, 3, 6, 7, 4, 0, 5]])
t2=tensor([[6, 4],
        [3, 0],
        [5, 5]])
t3=tensor([[5, 5],
        [3, 0],
        [6, 4]])
t4=tensor([[5, 3, 6],
        [5, 0, 4]])


In [ ]:
res = []

for in_vec in in_tensor:
  t1 = torch.argsort(in_vec, dim=-1)
  print(f"{t1=}")
  t2 = t1[-3:]
  print(f"{t2=}")
  t3 = t2.flipud()
  print(f"{t3=}")
  t4 = torch.argsort(in_vec, dim=-1)[-3:].flipud()
  # t4 = t3.T
  # print(f"{t4=}")
  res.append(t4)

res

t1=tensor([1, 2, 0, 7, 4, 6, 3, 5])
t2=tensor([6, 3, 5])
t3=tensor([5, 3, 6])
t1=tensor([1, 2, 3, 6, 7, 4, 0, 5])
t2=tensor([4, 0, 5])
t3=tensor([5, 0, 4])


[tensor([5, 3, 6]), tensor([5, 0, 4])]

In [ ]:
a1 = np.array([[1, 2, 3], [4, 5, 6]])
a2 = np.array([[1, 2, 3], [4, 5, 6]])
np.dot(a1, a2.T)

array([[14, 32],
       [32, 77]])

In [ ]:
torch.matmul(torch.tensor(a1), torch.tensor(a2).T)

tensor([[14, 32],
        [32, 77]])

In [ ]:
torch.cuda.empty_cache()

### CSLS Retrieval

In [ ]:
acc = compute_csls_accuracy(x_src, x_tgt, src2tgt, bsz=2048, top_k=1)
acc

0.3695797198132088

In [ ]:
acc = compute_csls_accuracy(x_src, x_tgt, src2tgt, bsz=2048, top_k=5)
acc

0.5690460306871248

In [ ]:
acc = compute_csls_accuracy(x_src, x_tgt, src2tgt, bsz=2048, top_k=10)
acc

0.6324216144096064

### NN Retrieval

In [ ]:
# acc = compute_nn_accuracy(x_src, x_tgt, src2tgt, bsz=100, top_k=10)
# 44min with creting the si_splits (wiki) -- <1s with created si_splits
acc = compute_nn_accuracy(x_src, x_tgt, src2tgt, bsz=100, top_k=10)
acc

100 done -- correct: 63.0
200 done -- correct: 129.0
300 done -- correct: 180.0
400 done -- correct: 241.0
500 done -- correct: 306.0
600 done -- correct: 369.0
700 done -- correct: 428.0
800 done -- correct: 485.0
900 done -- correct: 550.0
1000 done -- correct: 612.0
1100 done -- correct: 680.0
1200 done -- correct: 746.0
1300 done -- correct: 815.0
1400 done -- correct: 879.0
1500 done -- correct: 938.0
correct: 938.0


0.6257505003335557

In [ ]:
acc = compute_nn_accuracy(x_src, x_tgt, src2tgt, bsz=100, top_k=5)
acc

100 done -- correct: 56.0
200 done -- correct: 119.0
300 done -- correct: 166.0
400 done -- correct: 222.0
500 done -- correct: 284.0
600 done -- correct: 342.0
700 done -- correct: 397.0
800 done -- correct: 447.0
900 done -- correct: 506.0
1000 done -- correct: 561.0
1100 done -- correct: 626.0
1200 done -- correct: 690.0
1300 done -- correct: 754.0
1400 done -- correct: 814.0
1500 done -- correct: 871.0
correct: 871.0


0.5810540360240161

In [ ]:
acc = compute_nn_accuracy(x_src, x_tgt, src2tgt, bsz=100, top_k=1)
acc

100 done -- correct: 45.0
200 done -- correct: 93.0
300 done -- correct: 128.0
400 done -- correct: 170.0
500 done -- correct: 218.0
600 done -- correct: 269.0
700 done -- correct: 313.0
800 done -- correct: 356.0
900 done -- correct: 407.0
1000 done -- correct: 447.0
1100 done -- correct: 500.0
1200 done -- correct: 553.0
1300 done -- correct: 606.0
1400 done -- correct: 657.0
1500 done -- correct: 705.0
correct: 705.0


0.4703135423615744

#### en-tr (Wiki)
w/o prune: 0.365, 0.711, 0.785 </br>
with prune (do not count since test entries are ~600/1500): 0.410, 0.722, 0.807</br>

#### tr-en (Wiki)
0.640, 0.809, 0.843 </br>
--------------------------</br>
#### en-ja (Wiki)
w/o prune: 0.064, 0.415, 0.570 </br>
with prune: 0.378, 0.613, 0.688</br>

#### ja-en (Wiki)
0.267, 0.433, 0.492 </br>
--------------------------</br>
#### en-ta (Wiki)
w/o prune: 0.157, 0.571, 0.673 </br>
with prune: 0.362, 0.609, 0.672</br>

#### ta-en (Wiki)
0.357, 0.583, 0.659 </br>
--------------------------</br>
#### en-de (Wiki)
w/o prune: 0.453, 0.767, 0.829 </br>
with prune: </br>

#### de-en (Wiki)
0.56, 0.749, 0.786 </br>
--------------------------</br>

#### en-fr (Wiki)
w/o prune: 0.487, 0.861, 0.909 </br>
with prune: </br>

#### fr-en (Wiki)
0.724, 0.921, 0.941 </br>
--------------------------</br>

#### en-es (Wiki)
w/o prune: 0.400, 0.831, 0.896 </br>
with prune: </br>

#### es-en (Wiki)
0.653, 0.889, 0.921 </br>
--------------------------</br>

#### en-zh (Wiki)
w/o prune: 0.198, 0.447, 0.631</br>
with prune: 0.271, 0.648, 0.737</br>

#### zh-en (Wiki)
0.421, 0.634, 0.697</br>
--------------------------</br>
#### en-ru (Wiki)
w/o prune: 0.174, 0.651, 0.767</br>
with prune: 0.438, 0.693, 0.796</br>

#### ru-en (Wiki)
0.483, 0.783, 0.835</br>

#### en-it (Wiki)
w/o prune: 0.415, 0.791 0.852 </br>
with prune: </br>

#### it-en (Wiki)
0.645, 0.863, 0.903 </br>
--------------------------</br>

## LaBSE

### Wiki =======================
#### En-Si ---------------------
##### w/o prune En:
@10 - 0.7028647568287808</br>
@5 - 0.5842771485676216</br>
@1 - 0.04463690872751499</br>

##### with prune En:
@10 - 0.774</br>
@5 - 0.7086666666666667</br>
@1 -0.466</br>

##### with prune En and considering Si word splits (stems):
@10 - 0.8173333333333334</br>
@5 - 0.7713333333333333</br>
@1 - 0.578</br>

#### Si-En ---------------------
##### (ASCII based pruning is not needed for Si-En direction)
@10 - 0.7386666666666667</br>
@5 - 0.644</br>
@1 - 0.35933333333333334</br>

### CC =========================
#### En-Si ---------------------
##### w/o prune En:
@10 - 0.6473333333333333</br>
@5 - 0.5133333333333333 </br>
@1 - 0.03933333333333333 </br>

0.524
##### with prune En:
@10 - 0.7086666666666667 </br>
@5 - 0.6386666666666667 </br>
@1 - 0.36333333333333334 </br>

##### with prune En and considering Si word splits (stems):
@10 - 0.79 </br>
@5 - 0.7426666666666667 </br>
@1 - 0.524</br>

#### Si-En ---------------------
##### (ASCII based pruning is not needed for Si-En direction)
@10 - 0.7546666666666667 </br>
@5 - 0.696 </br>
@1 - 0.41 </br>

## XLM-R

### Wiki =======================
#### En-Si ---------------------
##### w/o prune En:
@10 - 0.14533333333333334 </br>
@5 - 0.09266666666666666 </br>
@1 - 0.004 </br>

##### with prune En:
@10 - 0.18266666666666667 </br>
@5 - 0.146 </br>
@1 - 0.06066666666666667 </br>

#### Si-En ---------------------
##### (ASCII based pruning is not needed for Si-En direction)
@10 - 0.14933333333333335 </br>
@5 - 0.10866666666666666 </br>
@1 - 0.034666666666666665 </br>

### CC =========================
#### En-Si ---------------------
##### w/o prune En:
@10 - 0.10933333333333334 </br>
@5 - 0.05733333333333333 </br>
@1 - 0.0013333333333333333 </br>

##### with prune En:
@10 - 0.158 </br>
@5 - 0.11733333333333333 </br>
@1 - 0.04666666666666667 </br>

#### Si-En ---------------------
##### (ASCII based pruning is not needed for Si-En direction)
@10 - 0.14466666666666667 </br>
@5 - 0.1 </br>
@1 - 0.008666666666666666 </br>